# Definition of ROS service LocRequestTimestamp for sick localization.
#
# ROS service LocRequestTimestamp requests a timestamp from the localization controller
# by sending cola command LocRequestTimestamp ("sMN LocRequestTimestamp").
#
# The service receives and decodes the current timestamp (uint32 timestamp in milliseconds)
# and calculates the time offset with the following formular:
#
# delta_time_ms := mean_time_vehicle_microsec - timestamp_lidar_microsec
# mean_time_vehicle_microsec := (send_time_vehicle + receive_time_vehicle) / 2
#                            := vehicles mean timestamp in microseconds
# send_time_vehicle          := vehicles timestamp when sending LocRequestTimestamp
# receive_time_vehicle       := vehicles timestamp when receiving the LocRequestTimestamp response
# timestamp_lidar_microsec   := lidar timestamp in microseconds from LocRequestTimestamp response
#
# See Operation-Instruction-v1.1.0.241R.pdf for details about time synchronization and
# time offset calculation. See Telegram-Listing-v1.1.0.241R.pdf and Technical_information_Telegram_Listing_NAV_LOC_en_IM0076556.pdf
# for further details about Cola telegram LocRequestTimestamp.

#
# Request (input)
#

---

#
# Response (output)
#

uint64 timestamp_lidar_microsec        # Lidar timestamp in microseconds from LocRequestTimestamp response
uint64 mean_time_vehicle_microsec      # Vehicle mean timestamp in microseconds: (send_time_vehicle + receive_time_vehicle) / 2
uint64 delta_time_microsec             # Time offset: mean_time_vehicle_microsec - timestamp_lidar_microsec
uint32 send_time_vehicle_sec           # Vehicle timestamp when sending LocRequestTimestamp (seconds part of ros timestamp immediately before tcp send)
uint32 send_time_vehicle_nsec          # Vehicle timestamp when sending LocRequestTimestamp (nano seconds part of ros timestamp immediately before tcp send)
uint32 receive_time_vehicle_sec        # Vehicle timestamp when receiving the LocRequestTimestamp response (seconds part of ros timestamp immediately after first response byte received)
uint32 receive_time_vehicle_nsec       # Vehicle timestamp when receiving the LocRequestTimestamp response (nano seconds part of ros timestamp immediately after first response byte received)
